mpl_interactions/controller.py

```diff
@@ -0,0 +1,187 @@
+
+    def slider_updated(self, change, key, values):
+        """
+        gotta also give the indices in order to support hyperslicer without horrifying contortions
+        """
+        with self.out:
+            if values is None:
+                self.params[key] = change["new"]
+            else:
+                self.params[key] = values[change["new"]]
+            self.indices[key] = change["new"]
+            for f, params in self._update_funcs[key]:
+                ps = {}
+                idxs = {}
+                for k in params:
+                    ps[k] = self.params[k]
+                    idxs[k] = self.indices[k]
+                f(params=ps, indices=idxs)
+            for f in self.figs[key]:
+                f.canvas.draw_idle()
+
+    def register_function(self, f, fig, params=None):
+        """
+        if params is None use the entire current set of params
+        """
+        if params is None:
+            params = self.params.keys()
+        # listify to ensure it's not a reference to dicts keys
+        # bc thats mutable
+        params = list(params)
+        for p in params:
+            self._update_funcs[p].append((f, params))
+            self.figs[p].append(fig)  # maybe should use a weakref?
+            # also should probably register a close_event callback to remove
+            # the figure
+
+    def __getitem__(self, key):
+        """
+        hack to allow calls like
+        interactive_plot(...beta=(0,1), controls = controls["tau"])
+        also allows [None] to grab None of the current params
+        to imply that we only want tau from the existing set of commands
+        """
+
+        # make sure keys is a list
+        # bc in gogogo_controls it may get added to another list
+        if isinstance(key, str):
+            key = [key]
+        elif key is None:
+            key = []
+        return self, key
+
+    def _ipython_display_(self):
+        ipy_display(self.vbox)
+
+
+def gogogo_controls(
+    kwargs, controls, display_controls, slider_formats, play_buttons, play_button_pos
+):
+    if controls:
+        if isinstance(controls, tuple):
+            # it was indexed by the user when passed in
+            extra_keys = controls[1]
+            controls = controls[0]
+            controls.add_kwargs(kwargs, slider_formats, play_buttons, play_button_pos)
+            params = {k: controls.params[k] for k in list(kwargs.keys()) + list(extra_keys)}
+        else:
+            controls.add_kwargs(kwargs, slider_formats, play_buttons, play_button_pos)
+            params = controls.params
+    else:
+        controls = Controls(
+            slider_formats=slider_formats,
+            play_buttons=play_buttons,
+            play_button_pos=play_button_pos,
+            **kwargs
+        )
+        params = controls.params
+        if display_controls:
+            if notebook_backend():
+                display(controls)
+            else:
+                pass
```

Comment:
in the mpl case should utilize a `controls.show` method that loops over the control figures and shows them all.

May make sense for controls to have both a `show` method and a `display` method so that it keeps the vocabulary of both ipywidgets and maptlotlib